File: 092_Reversed_Linked_List_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        
        newHead = ListNode(None)
        newHead.next = head
        lNode = newHead
        
        for num in range(1,m):
            lNode = lNode.next
        
        sNode = lNode.next
        for num in range(m,n):
            node = sNode.next
            sNode.next = node.next
            node.next = lNode.next
            lNode.next = node
        
        if m is 1:
            return lNode.next
        else:
            return newHead.next

File: 092_Reversed_Linked_List_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        
        ListNode *newHead = new ListNode(NULL);
        newHead->next = head;
        ListNode *lNode = newHead;
        
        for(int i=1; i<m; i++)
            lNode = lNode->next;
            
        ListNode *sNode = lNode->next;
        ListNode *node = new ListNode(NULL);
        for(int i=m; i<n; i++){
            node = sNode->next;
            sNode->next = node->next;
            node->next = lNode->next;
            lNode->next = node;
        }
        
        if(m==1)
            return lNode->next;
        else
            return head;
            
    }
};
